File: go/loader.go
package reflex

import (
	"encoding/json"
	"fmt"
)

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

// GuardRegistry maps guard names (from JSON) to Guard implementations.
type GuardRegistry map[string]Guard

// LoadWorkflowOptions configures optional parameters for LoadWorkflow.
type LoadWorkflowOptions struct {
	Guards GuardRegistry
}

// ---------------------------------------------------------------------------
// Intermediate JSON types (for edges with guards)
// ---------------------------------------------------------------------------

type jsonWorkflow struct {
	Schema   string             `json:"$schema"`
	ID       string             `json:"id"`
	Entry    string             `json:"entry"`
	Nodes    map[string]*Node   `json:"nodes"`
	Edges    []jsonEdge         `json:"edges"`
	Metadata map[string]any     `json:"metadata"`
}

type jsonEdge struct {
	ID    string           `json:"id"`
	From  string           `json:"from"`
	To    string           `json:"to"`
	Event string           `json:"event"`
	Guard json.RawMessage  `json:"guard,omitempty"`
}

// ---------------------------------------------------------------------------
// LoadWorkflow
// ---------------------------------------------------------------------------

// LoadWorkflow parses and validates a JSON workflow definition, returning a
// typed Workflow with guards resolved. Custom guard names are resolved against
// the GuardRegistry in opts.
func LoadWorkflow(data []byte, opts *LoadWorkflowOptions) (*Workflow, error) {
	var jw jsonWorkflow
	if err := json.Unmarshal(data, &jw); err != nil {
		return nil, newValidationError(ErrSchemaViolation, "<unknown>",
			fmt.Sprintf("invalid JSON: %v", err))
	}

	// Validate required top-level fields
	if jw.ID == "" {
		return nil, newValidationError(ErrSchemaViolation, "<unknown>",
			"missing required field: id")
	}
	if jw.Entry == "" {
		return nil, newValidationError(ErrSchemaViolation, jw.ID,
			"missing required field: entry")
	}
	if jw.Nodes == nil {
		return nil, newValidationError(ErrSchemaViolation, jw.ID,
			"missing required field: nodes")
	}
	if jw.Edges == nil {
		return nil, newValidationError(ErrSchemaViolation, jw.ID,
			"missing required field: edges")
	}

	// Validate nodes
	for key, node := range jw.Nodes {
		if node.ID == "" {
			return nil, newValidationError(ErrSchemaViolation, jw.ID,
				fmt.Sprintf("node '%s': missing required field: id", key))
		}
		if node.Spec == nil {
			return nil, newValidationError(ErrSchemaViolation, jw.ID,
				fmt.Sprintf("node '%s': missing required field: spec", key))
		}
	}

	// Process edges and resolve guards
	edges := make([]Edge, len(jw.Edges))
	for i, je := range jw.Edges {
		if je.ID == "" {
			return nil, newValidationError(ErrSchemaViolation, jw.ID,
				fmt.Sprintf("edge at index %d: missing required field: id", i))
		}
		if je.From == "" {
			return nil, newValidationError(ErrSchemaViolation, jw.ID,
				fmt.Sprintf("edge '%s': missing required field: from", je.ID))
		}
		if je.To == "" {
			return nil, newValidationError(ErrSchemaViolation, jw.ID,
				fmt.Sprintf("edge '%s': missing required field: to", je.ID))
		}
		if je.Event == "" {
			return nil, newValidationError(ErrSchemaViolation, jw.ID,
				fmt.Sprintf("edge '%s': missing required field: event", je.ID))
		}

		edge := Edge{
			ID:    je.ID,
			From:  je.From,
			To:    je.To,
			Event: je.Event,
		}

		if len(je.Guard) > 0 {
			guard, err := resolveGuardJSON(je.Guard, jw.ID, je.ID, opts)
			if err != nil {
				return nil, err
			}
			edge.Guard = guard
		}

		edges[i] = edge
	}

	return &Workflow{
		ID:       jw.ID,
		Entry:    jw.Entry,
		Nodes:    jw.Nodes,
		Edges:    edges,
		Metadata: jw.Metadata,
	}, nil
}

// ---------------------------------------------------------------------------
// Guard resolution
// ---------------------------------------------------------------------------

func resolveGuardJSON(raw json.RawMessage, wfID, edgeID string, opts *LoadWorkflowOptions) (Guard, error) {
	var gm map[string]any
	if err := json.Unmarshal(raw, &gm); err != nil {
		return nil, newValidationError(ErrSchemaViolation, wfID,
			fmt.Sprintf("edge '%s': invalid guard JSON: %v", edgeID, err))
	}

	typ, ok := gm["type"].(string)
	if !ok {
		return nil, newValidationError(ErrSchemaViolation, wfID,
			fmt.Sprintf("edge '%s': guard missing or invalid 'type' field", edgeID))
	}

	switch typ {
	case "exists", "not-exists":
		key, ok := gm["key"].(string)
		if !ok {
			return nil, newValidationError(ErrSchemaViolation, wfID,
				fmt.Sprintf("edge '%s': guard type '%s' requires 'key' field", edgeID, typ))
		}
		return &BuiltinGuard{Type: GuardType(typ), Key: key}, nil

	case "equals", "not-equals":
		key, ok := gm["key"].(string)
		if !ok {
			return nil, newValidationError(ErrSchemaViolation, wfID,
				fmt.Sprintf("edge '%s': guard type '%s' requires 'key' field", edgeID, typ))
		}
		// value is required but may be null
		value, hasValue := gm["value"]
		if !hasValue {
			return nil, newValidationError(ErrSchemaViolation, wfID,
				fmt.Sprintf("edge '%s': guard type '%s' requires 'value' field", edgeID, typ))
		}
		return &BuiltinGuard{Type: GuardType(typ), Key: key, Value: value}, nil

	case "custom":
		name, ok := gm["name"].(string)
		if !ok {
			return nil, newValidationError(ErrSchemaViolation, wfID,
				fmt.Sprintf("edge '%s': custom guard requires 'name' field", edgeID))
		}
		if opts == nil || opts.Guards == nil || opts.Guards[name] == nil {
			return nil, newValidationError(ErrUnknownGuardReference, wfID,
				fmt.Sprintf("edge '%s': custom guard '%s' not found in guard registry", edgeID, name))
		}
		return opts.Guards[name], nil

	default:
		return nil, newValidationError(ErrSchemaViolation, wfID,
			fmt.Sprintf("edge '%s': unknown guard type '%s'", edgeID, typ))
	}
}

File: typescript/src/loader.ts
// Reflex — Workflow Loader
// Implements M7-2: Load and validate JSON workflow definitions

import Ajv from 'ajv';
import type {
  BlackboardReader,
  BuiltinGuard,
  Edge,
  Guard,
  InvocationSpec,
  Node,
  NodeSpec,
  Workflow,
} from './types.js';
import { WorkflowValidationError } from './registry.js';
import { workflowSchema } from './workflow-schema.js';

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

/** Maps guard names (from JSON) to evaluate functions. */
export type GuardRegistry = Record<
  string,
  (blackboard: BlackboardReader) => boolean
>;

/** Options for loadWorkflow. */
export interface LoadWorkflowOptions {
  guards?: GuardRegistry;
}

// ---------------------------------------------------------------------------
// Schema validator (compiled once at module load)
// ---------------------------------------------------------------------------

const ajv = new Ajv({ allErrors: true });
const validateSchema = ajv.compile(workflowSchema);

// ---------------------------------------------------------------------------
// Guard shapes from JSON (before resolution)
// ---------------------------------------------------------------------------

interface JsonBuiltinGuard {
  type: 'exists' | 'not-exists' | 'equals' | 'not-equals';
  key: string;
  value?: unknown;
}

interface JsonCustomGuard {
  type: 'custom';
  name: string;
}

type JsonGuard = JsonBuiltinGuard | JsonCustomGuard;

// ---------------------------------------------------------------------------
// loadWorkflow
// ---------------------------------------------------------------------------

/**
 * Parse and validate a JSON workflow definition, returning a typed Workflow.
 *
 * @param input - Raw JSON string or pre-parsed object
 * @param options - Optional guard registry for resolving custom guard references
 * @returns A fully typed Workflow with guards resolved
 * @throws WorkflowValidationError on schema violations or missing guard references
 */
export function loadWorkflow(
  input: unknown,
  options?: LoadWorkflowOptions,
): Workflow {
  // 1. Parse if string
  let data: unknown;
  if (typeof input === 'string') {
    try {
      data = JSON.parse(input);
    } catch (e) {
      throw new WorkflowValidationError(
        'SCHEMA_VIOLATION',
        '<unknown>',
        `Invalid JSON: ${(e as Error).message}`,
      );
    }
  } else {
    data = input;
  }

  // 2. Validate against schema
  if (!validateSchema(data)) {
    const errors = validateSchema.errors ?? [];
    const detail = errors
      .map((e) => `${e.instancePath || '/'}: ${e.message}`)
      .join('; ');
    throw new WorkflowValidationError(
      'SCHEMA_VIOLATION',
      '<unknown>',
      `Schema validation failed: ${detail}`,
      { errors },
    );
  }

  // From here the data is structurally valid per the schema.
  const raw = data as Record<string, unknown>;

  const id = raw.id as string;
  const entry = raw.entry as string;
  const rawNodes = raw.nodes as Record<string, Record<string, unknown>>;
  const rawEdges = raw.edges as Array<Record<string, unknown>>;
  const metadata = raw.metadata as Record<string, unknown> | undefined;

  // 3. Build nodes
  const nodes: Record<string, Node> = {};
  for (const [key, rawNode] of Object.entries(rawNodes)) {
    const node: Node = {
      id: rawNode.id as string,
      spec: (rawNode.spec ?? {}) as NodeSpec,
    };
    if (rawNode.description !== undefined) {
      node.description = rawNode.description as string;
    }
    if (rawNode.invokes !== undefined) {
      const inv = rawNode.invokes as Record<string, unknown>;
      node.invokes = {
        workflowId: inv.workflowId as string,
        returnMap: (inv.returnMap as Array<Record<string, string>>).map(
          (rm) => ({
            parentKey: rm.parentKey,
            childKey: rm.childKey,
          }),
        ),
      } satisfies InvocationSpec;
    }
    nodes[key] = node;
  }

  // 4. Build edges with guard resolution
  const edges: Edge[] = rawEdges.map((rawEdge) => {
    const edge: Edge = {
      id: rawEdge.id as string,
      from: rawEdge.from as string,
      to: rawEdge.to as string,
      event: rawEdge.event as string,
    };

    if (rawEdge.guard !== undefined) {
      edge.guard = resolveGuard(
        rawEdge.guard as JsonGuard,
        id,
        edge.id,
        options?.guards,
      );
    }

    return edge;
  });

  // 5. Assemble workflow
  const workflow: Workflow = { id, entry, nodes, edges };
  if (metadata !== undefined) {
    workflow.metadata = metadata;
  }

  return workflow;
}

// ---------------------------------------------------------------------------
// Guard resolution
// ---------------------------------------------------------------------------

function resolveGuard(
  json: JsonGuard,
  workflowId: string,
  edgeId: string,
  registry?: GuardRegistry,
): Guard {
  if (json.type === 'custom') {
    const fn = registry?.[json.name];
    if (!fn) {
      throw new WorkflowValidationError(
        'UNKNOWN_GUARD_REFERENCE',
        workflowId,
        `Edge '${edgeId}': custom guard '${json.name}' not found in guard registry`,
        { edgeId, guardName: json.name },
      );
    }
    return { type: 'custom', evaluate: fn };
  }

  // Builtin guard — pass through directly
  const guard: BuiltinGuard = { type: json.type, key: json.key };
  if ('value' in json && json.value !== undefined) {
    guard.value = json.value;
  }
  return guard;
}
